libobs/obs-encoder.c

```diff
@@ -810,34 +812,26 @@ static inline bool obs_encoder_stop_internal(
 
 	if (last) {
 		remove_connection(encoder, true);
-		encoder->initialized = false;
+		pthread_mutex_unlock(&encoder->init_mutex);
 
-		if (encoder->destroy_on_stop) {
-			pthread_mutex_unlock(&encoder->init_mutex);
-			obs_encoder_actually_destroy(encoder);
-			return true;
-		}
-	}
+		struct obs_encoder_group *group = encoder->encoder_group;
 
-	return false;
-}
+		if (encoder->destroy_on_stop)
+			obs_encoder_actually_destroy(encoder);
 
-void obs_encoder_stop(obs_encoder_t *encoder,
-		      void (*new_packet)(void *param,
-					 struct encoder_packet *packet),
-		      void *param)
-{
-	bool destroyed;
+		/* destroying the group all the way back here prevents a race where destruction
+		 * of the group can prematurely destroy the encoder within internal functions.
+		 * this is the point where it is safe to destroy the group, even if the encoder
+		 * is then also destroyed */
+		if (group && group->destroy_on_stop &&
+		    group->num_encoders_started == 0)
+			obs_encoder_group_actually_destroy(group);
```

Comment:
```suggestion
		if (!group)
			return;
		
		pthread_mutex_lock(&group->mutex);
		if (group->destroy_on_stop &&
		    group->num_encoders_started == 0) {
			obs_encoder_group_actually_destroy(group);
		} else {
			pthread_mutex_unlock(&group->mutex);
		}
```

`obs_encoder_group_actually_destroy` currently unlocks the mutex before destroying it, so the mutex needs to be locked before calling `obs_encoder_group_actually_destroy`; additionally, accesses to group members are protected by the mutex